woocommerce/woocommerce-gutenberg-products-block
# Expand Feature flags

closes #2451 

This PR changes and expand our feature flags to make them more dynamic and to work with the core/plugin/dev model, so it does the following:

- It changes our flags names from `stable` and `experimental` to numbers, this is a requirement so we can have collapsable flags, so we have 3 flags:
    - `experimental` is now `3`.
    - `plugin` is now `2`.
    - `core` is now `1`.
 for most cases you always compare up, this means if you want something to work only in dev you check if the flag is bigger than 2 `WOOCOMMERCE_BLOCKS_PHASE > 2`, if you want something to be in the feature plugin you compare for bigger than 1 `WOOCOMMERCE_BLOCKS_PHASE > 1`.

- It updates our build process slightly, core and stable files and blocks should live `stableMainEntry` and `stableFrontEndEntry`, this means things like Cart, Checkout, and other blocks, new unstable blocks like Single Product should live in `experimentalMainEntry` or `experimentalFrontEndEntry`, those files are only built if you run `npm run build` and `npm start`, they won't be built if you run `npm run deploy`, you can force not building them locally by passing a different flag so `WOOCOMMERCE_BLOCKS_PHASE=2 npm run build`.

- There's now a constant named `WC_BLOCKS_IS_FEATURE_PLUGIN`, it should be defined if we're in running feature plugin and won't be defined when running from core, in most cases, you won't use this constant.

- in PHP you will still use `WOOCOMMERCE_BLOCKS_PHASE` to define your features, same as PHP with same exact flags.

- In JS, you won’t be using `process.env.WOOCOMMERCE_BLOCKS_PHASE` anymore, you will have to use `WOOCOMMERCE_BLOCKS_PHASE` a constant exported from blocks settings

```js
import { WOOCOMMERCE_BLOCKS_PHASE } from '@woocommerce/block-settings';

if ( WOOCOMMERCE_BLOCKS_PHASE > 1 ) {
	registerBlockType( 'woocommerce/checkout', settings );
}
if ( WOOCOMMERCE_BLOCKS_PHASE > 2 ) {
	registerBlockType( 'woocommerce/single-product', settings );
}
```

- `woocommerce/feature-flag` eslint rule is disabled until we merge this and update it.

- Our deployment command `npm run deploy` will now append `WOOCOMMERCE_BLOCKS_PHASE=2` this means when you deploy, it will work on the context of feature plugin, but WooCommrce core will consume it on the context of Core automatically.

## How should you gate your code?

In most cases, you will have code that is:
- safe to run in WooCommerce core, do nothing.
- safe to run in the feature plugin only, it will be either:
    -  A file, wrap the enqueue of that file in a flag.
    - A code block, wrap it in `WOOCOMMERCE_BLOCKS_PHASE`.
- A code that only runs in development or `experimental`, it will be either:
    - A file, wrap the enqueue, this is the most important part.
    - Skip building the file (to reduce the final zip file) by moving it to `experimentalMainEntry` or `experimentalFrontEndEntry`.
   - A code block, wrap it in `WOOCOMMERCE_BLOCKS_PHASE`.
   - A JS code block, if it’s heavy and will affect the final bundle size, we can see into reintroducing dead code elimination (I didn't add this because I wanted to reduce complexity unless we have a use case for it).


## Test instructions:

- Build the plugin normally `npm run build`, you should be able to add any block, including Cart and Checkout and Single Product.

- Build the plugin in feature plugin mode `WOOCOMMERCE_BLOCKS_PHASE=2 npm run build`, you should be able to add any block, including Cart and Checkout but **not Single Product.**

- Build plugin in Core mode `WOOCOMMERCE_BLOCKS_PHASE=1 npm run build`, you should be able to add our core blocks, and **not Cart, Checkout, or Single Product.**

- If you want to further test, you can try to create a zip `WOOCOMMERCE_BLOCKS_PHASE=2 npm run package-plugin` to generate a zip that you can upload to other websites, it would simulate a feature plugin.
